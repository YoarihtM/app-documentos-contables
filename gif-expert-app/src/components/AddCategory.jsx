import { useState } from 'react';
import { PropTypes } from 'prop-types';

// No es comun ver las props, por lo regular se desestructura y
// se recibe las props que requerimos
// export const AddCategory = ( props ) => {

export const AddCategory = ({ onNewCategory }) => {
    const [inputValue, setInputValue] = useState('');

    const onInputChange = ({ target }) => {

        // console.log(target.value);
        setInputValue(target.value);

    };

    const onSubmit = ( event ) => {
        // console.log('Hola mundo desde el submit de AddCategory');
        // console.log(event);

        // Mi intento de la tarea
        // console.log(props.setCategories);
        // console.log(props.categories);
        // console.log(event.target[0].value);
        
        // props.setCategories([ ...props.categories, event.target[0].value]);
        // Fin del intento, no funciono


        // console.log(inputValue);
        
        // Solucion 1 del profe junto a validaciones

        // if( inputValue.trim().length <= 1 ) return;

        // setCategories( categories => [ ...categories, inputValue ]);
        // setInputValue('');
        
        // Solucion 2 del profe
        
        event.preventDefault();
        const newInputValue = inputValue.trim();
        
        if(newInputValue <= 1) return;
        
        onNewCategory(newInputValue);
        setInputValue('');

    };

    return (
        <form onSubmit={ onSubmit } aria-label="form">
            <input
                type="text"
                placeholder="Buscar gifs"
                value={inputValue}
                onChange={onInputChange}
            />
        </form>
    )
}


AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired,
};