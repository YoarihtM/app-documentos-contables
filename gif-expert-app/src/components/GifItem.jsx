

export const GifItem = () => {
    return (
        <div>GifItem</div>
    )
}
