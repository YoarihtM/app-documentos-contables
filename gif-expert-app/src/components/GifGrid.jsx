// El orden de importación siempre va hasta arriba la funcionalidad
// de la biblioteca o framework, seguido de los componentes,
// seguido de los hooks y por último los helpers

import { useEffect, useState } from 'react';
import { PropTypes } from 'prop-types';

import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({ category }) => {

    // Se usa el estado para guardar las imagenes de cada busqueda 
    // y para mostrar los titulos de las imagenes resultantes
    // ese mismo estado se usará dentro del useEffect para realizar
    // los cambios de cada busqueda y hay dos formas de hacerlo
    // const [images, setImages] = useState([]);

    // el hook useEffect sirve para disparar "Efectos secundarios"
    // luego de que se realiza una tarea
    // Tiene dos argumentos el primero es una funcion que ejecuta
    // el codigo que queremos disparar y el segundo es una lista o
    // arreglo que indica cuando se va a ejecutar el primer parametro
    // si se deja vacio significa que solo se va a ejecutar la primera
    // vez que se renderice el componente
    
    // useEffect( () => {
    //     getGifs(category);
    // }, [ ])

    // La primera forma de llamar las imagenes con useEffect
    // aqui se usa el then ya que useEffect no acepta en el 
    // primer parametro una promesa y siempre debe ser una 
    // funcion, asi se mantiene sin el uso de async/await
    
    // useEffect( () => {
    //     getGifs(category)
    //         .then( newImages => setImages(newImages));
    // }, [ ])

    // La segunda forma es usando async/await con una funcion externa
    // const getImages = async () => {
    //     const newImages = await getGifs(category);

    //     setImages(newImages);
    // }

    // y al usar el hook useEffect ya se puede llamar la funcion
    // en el primer parametro, siendo ahora si permitido
    // useEffect( () => {
    //     getImages();
    // }, [ ])


    //////////////////////////////////////////////////////////////////////

    // En esta nueva sección se va a realizar un custom hook con lo cual
    // todo el código que está arriba va a dejar de usarse y se reemplazará
    // por el custom hook que va a realizar toda la lógica que lleva a cabo
    // dicho código, con este custom hook se busca implementar DRY ya que
    // es parte del principio el poder usar ese hook en otros componentes

    const { images, isLoading } = useFetchGifs(category);

    return (

        <>
            <h3>{category}</h3>

            { 
                // isLoading ? <h2>Cargando...</h2> : null 
                isLoading && (<h2>Cargando...</h2>)
            }

            <div className="card-grid">
                {
                    images.map( (image) => (
                        // <li key={id}>{ title }</li>
                        <GifItem 
                            key={image.id} 
                            // title={image.title} 
                            // url={image.url} 
                            { ...image }
                        />
                    ))
                }
            </div>



        </>

    )
}


GifGrid.propTypes = {
    category: PropTypes.string.isRequired

};