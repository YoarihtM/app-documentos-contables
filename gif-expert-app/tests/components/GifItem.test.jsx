import { render } from '@testing-library/react';
import { GifItem } from '../../src/components/GifItem';

describe('Pruebas de GifItem', () => { 
    
            const title = 'titulo';
            const url   = 'url';

    test('debe de hacer match con el snapshot', () => {

        const { container } = render(<GifItem title={ title } url={ url } />)
        expect(container).toMatchSnapshot();

    });

    test('Debe de mostrar la imagen con el URL y el ALT indicado', () => {  
        
        render(
            <GifItem title={title} url={url} />
        );

        // screen.debug();

        // expect(screen.getByRole('img').src).toBe( url );
        // expect(screen.getByRole('img').alt).toBe( title );

        const { src, alt} = screen.getByRole('img');
        expect(src).toBe(url);
        expect(alt).toBe(title);

    });

}); 