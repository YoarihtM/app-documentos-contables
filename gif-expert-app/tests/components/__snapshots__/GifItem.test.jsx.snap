// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de GifItem debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitama"
      src="https://one-punch.com/saitama.jpg"
    />
    <p>
      Saitama
    </p>
  </div>
</div>
`;
